scripts/hud/synchronizer.js

```diff
@@ -31,32 +31,58 @@ class Synchronizer {
 	static firstPanelWidth = this.maxSegmentWidth;
 	static syncGain = 10; // scale how fast the bars move
 	static altColor = 'rgba(0, 0, 0, 0.0)';
+	static lastVel = { x: 0, y: 0 };
+	static lastYaw = 0;
 
 	static onLoad() {
 		this.initializeSettings();
 
 		if (this.mom_hud_synchro_stat_mode) this.onJump(); // show stats if enabled
+
+		this.airSpeed = 30;
+		this.runSpeed = GameInterfaceAPI.GetSettingFloat('sv_maxspeed');
+		this.acceleration = GameInterfaceAPI.GetSettingFloat('sv_accelerate');
+		this.airAcceleration = GameInterfaceAPI.GetSettingFloat('sv_airaccelerate');
+		this.stopSpeed = GameInterfaceAPI.GetSettingFloat('sv_stopspeed');
+		this.friction = GameInterfaceAPI.GetSettingFloat('sv_friction');
+		this.tickInterval = MomentumMovementAPI.GetTickInterval();
+		this.maxGroundSpeed = this.getMaxGroundSpeed();
 	}
 
 	static onUpdate() {
 		const lastMoveData = MomentumMovementAPI.GetLastMoveData();
 		const lastTickStats = MomentumMovementAPI.GetLastTickStats();
 		const wishDir = lastMoveData.wishdir;
-		const velDir = this.getNormal(MomentumPlayerAPI.GetVelocity(), 0.001);
+
+		const vel = MomentumPlayerAPI.GetVelocity();
+		const velDir = this.getNormal(vel, 0.001);
+		const lastSpeed = this.getSize(this.lastVel);
+		this.lastVel = vel;
 
 		const bValidWishMove = this.getSize(wishDir) * this.getSize(velDir) > 0.001;
 		const strafeRight = (bValidWishMove ? 1 : 0) * lastTickStats.strafeRight;
 		const direction = this.mom_hud_synchro_dynamic_enable === 1 ? strafeRight : 1;
 		const flip = this.mom_hud_synchro_flip_enable === 1 ? -1 : 1;
 
+		const yaw = MomentumPlayerAPI.GetAngles().y;
+		const yawDelta = bValidWishMove ? this.wrapValueToRange(this.lastYaw - yaw, -180, 180) * this.deg2rad : 0;
+		this.lastYaw = yaw;
+
+		lastMoveData.isGrounded = lastMoveData.acceleration !== this.airAcceleration; // lastMoveData.moveStatus doesn't work
+		lastMoveData.maxspeed = lastMoveData.isGrounded ? this.runSpeed : this.airSpeed;
+
+		const spd = lastMoveData.isGrounded ? this.applyFriction(this.maxGroundSpeed) : lastSpeed;
+		const optimalDelta = this.findOptimalDelta(spd, lastMoveData);
+
 		this.addToBuffer(
 			this.gainRatioHistory,
 			this.sampleWeight * this.NaNCheck(lastTickStats.speedGain / lastTickStats.idealGain, 0)
 		);
 		const gainRatio = this.getBufferedSum(this.gainRatioHistory);
 
-		const ratio = this.mom_hud_synchro_mode > 2 ? 1 - lastTickStats.yawRatio : lastTickStats.yawRatio;
-		this.addToBuffer(this.yawRatioHistory, this.sampleWeight * this.NaNCheck(ratio, 0));
+		const ratio =
+			this.mom_hud_synchro_mode > 2 ? 1 - Math.abs(yawDelta) / optimalDelta : Math.abs(yawDelta) / optimalDelta;
+		this.addToBuffer(this.yawRatioHistory, this.sampleWeight * this.FiniteCheck(ratio, 0));
```

Comment:
Since `finiteCheck` is so similar to `Number.isFinite` and only being used once, I don't think there's much point defining a separate function. So can just do `this.sampleWeight * (Number.isFinite(ratio) ? ratio : 0)`